cips/cip-31.md

```diff
@@ -0,0 +1,157 @@
+| discussions-to |                                                                                                                         |
+| status         | Draft                                                                                                                   |
+| type           | Standards Track                                                                                                         |
+| category       | Core                                                                                                                    |
+| created        | 2025-02-07                                                                                                              |
+| requires       | N/A                                                                                                                     |
+
+---
+
+## Abstract
+
+*Note: in the software lockup accounts are referred to vesting accounts, for this document we will refer to them as lockup accounts.*
+
+This CIP proposes integrating staking rewards directly into lockup accounts. When a lockup account receives rewards from staking, the system adds the reward amount to the lockup balance and recalculates the daily unlock rate over the remaining lockup period. This mechanism ensures that reward tokens remain subject to an account’s existing lockup constraints while still providing incentive to secure the network through staking.
+
+## Motivation
+
+The default design of the lockup accounts in most Proof-of-Stake blockchains is that locked tokens can be staked and earn rewards, but these rewards are not subject to the lockup schedule. This has led to recipients of locked tokens being able to sell staking rewards independently of their unlock schedule, arguably in tension with the purpose of lockup accounts.
+
+By integrating staking rewards directly into the lockup schedule, this proposal incentivizes locked tokens to secure the network while ensuring that staking rewards are locked. Summarizing the motivations:
+
+- **Consistency:** Ensures that all tokens under a lockup account—whether from the initial locked amount or subsequently earned rewards—are subject to the same unlock schedule.  
+- **Security:** Prevents immediate liquidity from staking rewards without removing incentive to secure the network.  
+- **Simplicity:** Leverages existing modules (distribution, lockup, bank) with minimal disruption by adding hooks and type checks.
+
+## Specification
+
+### Core Mechanism
+
+1. **Account Type Verification:**  
+   When staking rewards are distributed (e.g., via `MsgWithdrawDelegatorReward`), the distribution module will first check if the receiving account is a lockup account.  
+
+   - **If it is:**  
+     - Verify that it is of a modifiable lockup type.  
+     - Calculate the new lockup amount:  
+       `New Lockup Amount = Original Lockup Amount + Claimed Rewards`  
```

Comment:
[question] In this context is "Claimed Rewards" only applicable to the claimed rewards at a particular moment in time? Put another way, this CIP only impacts claimed rewards _after_ this feature has been added. Lockup accounts that have already claimed staking rewards (i.e. on celestia-app v3) don't contribute to the "Claimed Rewards" in this calculation.

Reply:
correct, claims prior to the upgrade are not subject to locking at this would require large amount of offchain data